odl/operator/operator.py

```diff
@@ -95,57 +80,279 @@ def _default_apply(self, x, out, *args, **kwargs):
+        kw_only_defaults = {}
 
-    def __call__(cls, *args, **kwargs):
-        """Create a new class ``cls`` from given arguments.
+    signature = _signature_from_spec(call)
 
-        Raises
-        ------
-        NotImplementedError
-            If neither ``_call`` or ``_apply`` is implemented.
-        """
-        obj = type.__call__(cls, *args, **kwargs)
-        if not hasattr(obj, '_call') and not hasattr(obj, '_apply'):
-            raise NotImplementedError('`Operator` instances must'
-                                      'either have `_call` or `_apply`'
-                                      ' as attribute.')
-        return obj
+    pos_args = spec.args
+    if unbound_call is not None:
+        # Add 'self' to positional arg list to satisfy the checker
+        pos_args.insert(0, 'self')
+
+    pos_defaults = spec.defaults
+    varargs = spec.varargs
+
+    out_optional = False
+
+    # Variable args are not allowed
+    if varargs is not None:
+        raise ValueError("Bad signature '{}': variable arguments not allowed."
+                         " ".format(signature) + spec_msg)
 
+    if len(pos_args) not in (2, 3):
+        raise ValueError("Bad signature '{}'. ".format(signature) + spec_msg)
+
+    true_pos_args = pos_args[1:]
+    if len(true_pos_args) == 1:  # 'out' kw-only
+        if 'out' in true_pos_args:  # 'out' positional and 'x' kw-only -> no
+            raise ValueError("Bad signature '{}': `out` cannot be the only "
+                             "positional argument."
+                             " ".format(signature) + spec_msg)
+        else:
+            if 'out' not in kw_only:
+                has_out = False
+            elif kw_only_defaults['out'] is not None:
+                raise ValueError(
+                    "Bad signature '{}': `out` can only default to "
+                    "`None`, got '{}'."
+                    " ".format(signature, kw_only_defaults['out']) +
+                    spec_msg)
+            else:
+                has_out = True
+                out_optional = True
+
+    elif len(true_pos_args) == 2:  # Both args positional
+        if true_pos_args[0] == 'out':  # out must come second
+            py3_txt = ' or keyword-only. ' if py3 else '. '
+            raise ValueError("Bad signature '{}': `out` can only be the "
+                             "second positional argument".format(signature) +
+                             py3_txt + spec_msg)
+        elif true_pos_args[1] != 'out':  # 'out' must be 'out'
+            raise ValueError("Bad signature '{}': output parameter must "
+                             "be called 'out', got '{}'."
+                             " ".format(signature, true_pos_args[1]) +
+                             spec_msg)
+        else:
+            has_out = True
+            out_optional = bool(pos_defaults)
+            if pos_defaults and pos_defaults[-1] is not None:
+                raise ValueError("Bad signature '{}': `out` can only "
```

Comment:
The reason is that it does not translate to the public interface otherwise, since `__call__` is fixed to default to `None`, so any other value in `_call` wouldn't propagate. We can allow it, of course.